### PR TITLE
[WIP] Add support for installing and build dependencies, install folly using externalproject add

### PR DESCRIPTION
**Pull request by raulcd**
This PR is WIP.

This PR creates a macro to allow us to download/build/install dependencies using ExternalProject_ADD.
It is being tested with Folly.

Currently this PR is able to build and run all velox tests successfully downloading and installing folly on build time. To reproduce use:
```
docker-compose build --no-cache ubuntu-cpp
docker-compose run --rm ubuntu-cpp
```
TODO list:
- [x] Fix usage of CMAKE_CXX_FLAGS instead of hardcoding
- [x] Fix transient dependencies to not require to define all Folly dependencies via INTERFACE_LINK_LIBRARIES for transitive link dependencies
- [ ] Add CI to test both using the system installed dependency and downloading dependency via CMAKE
- [ ] Test on other OS not only docker

### CODE REVIEW

**Comment by raulcd** [comment]
@kgpai @assignUser See initial PR for what we have been discussing. The approach on this PR uses ExternalProject_Add for the third party dependencies. Currently this PR downloads and builds folly as part of the build and all tests pass on Docker. There are still several TO DO items but I wanted to share the approach.

File: ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+add_custom_target(toolchain)
+include(ExternalProject)
+
+#=====================================FOLLY==============================================
+
+if(DEFINED ENV{VELOX_FOLLY_URL})
+  set(FOLLY_SOURCE_URL "$ENV{VELOX_FOLLY_URL}")
+else()
+  set(VELOX_FOLLY_BUILD_VERSION v2022.07.11.00)
```
Comment (kgpai) [inline ThirdpartyToolchain.cmake:47]: Nit: We can probably do this in another PR, but would be nice if we have all our dependencies defined in another file (akin to how python dependencies are configured) and pull that in.

File: ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  set(FOLLY_STATIC_LIB
+      "${FOLLY_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}folly${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  set(FOLLY_BENCHMARK_STATIC_LIB
+      "${FOLLY_PREFIX}/lib/${CMAKE_STATIC_LIBRARY_PREFIX}follybenchmark${CMAKE_STATIC_LIBRARY_SUFFIX}"
+  )
+  # TODO Correctly use CMAKE_CXX_FLAGS defined on the project. Using them makes folly linking
+  # issues as described on:
+  # https://github.com/facebook/folly/blob/main/folly/container/detail/F14IntrinsicsAvailability.h#L32-L34
+  set(CMAKE_CXX_FLAGS_FOLLY "-mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17")
```
Comment (kgpai) [inline ThirdpartyToolchain.cmake:68]: These flags should be pulled from setup-helper-functions.sh; See here for an example on how to pull these in : https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L168 .

File: ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+      "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS_FOLLY}"
+      -DCMAKE_CXX_STANDARD=17
+      -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+      "-DCMAKE_INSTALL_PREFIX=${FOLLY_PREFIX}")
+
+  externalproject_add(folly
+                      URL ${FOLLY_SOURCE_URL}
+                      URL_HASH "SHA256=${VELOX_FOLLY_BUILD_SHA256_CHECKSUM}"
+                      CMAKE_ARGS ${FOLLY_CMAKE_ARGS}
+                      BUILD_BYPRODUCTS "${FOLLY_STATIC_LIB}" "${FOLLY_BENCHMARK_STATIC_LIB}")
```
Comment (kgpai) [inline ThirdpartyToolchain.cmake:79]: curious: do we explicitly need to ask it to build the static and benchmark lib ?

File: ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+macro(build_dependency DEPENDENCY_NAME)
+  if("${DEPENDENCY_NAME}" STREQUAL "folly")
+    build_folly()
+  else()
+    message(FATAL_ERROR "Unknown thirdparty dependency to build: ${DEPENDENCY_NAME}")
+  endif()
+endmacro()
+
+macro(resolve_dependency DEPENDENCY_NAME)
```
Comment (kgpai) [inline ThirdpartyToolchain.cmake:179]: nit: any reason why macro over function ? Also can we add some documentation here.

Comment (raulcd) [reply ThirdpartyToolchain.cmake:179]: The initial reason is that we use macros on ARROW. We can change it to functions, we would require some minor changes as some variables would be outside of the scope and they are required to be accessible, i.e:
```
set(FOLLY_LIBRARIES Folly::folly CACHE INTERNAL "Folly dependencies libraries")
```

File: ThirdpartyToolchain.cmake (hunk 3 above)
Comment (raulcd) [reply ThirdpartyToolchain.cmake:79]: At the moment we do. I can investigate a little further around it but if we don't declare the byproducts to build explicitly we get the following error:
```
ninja: error: 'folly-install/lib/libfolly.a', needed by 'velox/buffer/tests/velox_buffer_test', missing and no known rule to make it
```

File: ThirdpartyToolchain.cmake (hunk 4 above)
Comment (kgpai) [reply ThirdpartyToolchain.cmake:179]: I have no strong opinions on macros vs functions here; I am curious why ARROW chose that , are there any benefits you folks saw ?

Comment (raulcd) [reply ThirdpartyToolchain.cmake:179]: I've found the original discussion when it was implemented about the question on why use macro over function: https://github.com/apache/arrow/pull/3688/files#r262737838
I don't think there was another benefit apart from helping exposing some variables globally due to the scope difference.

File: ThirdpartyToolchain.cmake (hunk 2 above)
Comment (raulcd) [reply ThirdpartyToolchain.cmake]: Thanks! Done!

**Pull request closed without merging**